### PR TITLE
Feature/testing recipe

### PR DESCRIPTION
**Pull request by horaklukas**
Introduction and first chapter of "Testing our apps" series

### CODE REVIEW

**Comment by baierjak** [comment]
I think it would be really good to put there the comments with the result of the test. As it is in those tests at the begining -> right now some oh those are missing and it could be confusing for people when they dont know the final result

File: pages/testing/Essentials.md [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+import { selectMenuItems } from '../mySelectors.js'
+import { routingSelector } from '@ackee/chris'
+
+jest.mock('@ackee/chris')
+
+describe('selectMenuItems', () => {
+    it('returns all menu items', () => {
+        routingSelector.mock.returnValue({ pathname: '/' })
+        
+        const items = selectMenuItems({})
```
Comment (baierjak) [inline pages/testing/Essentials.md]: useless brackets

File: pages/Testing.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Automated testing
+
+We want working apps. But how can we be sure they old features works when we just refactored its code because of a new feature? We need to test the old feature, but how?
+
+Sure we can manually click on all buttons, fill out fields, open modals and visually check it works. But it's slow and imagine you should do it for every important part 
+of the app every week.   
+
+Here comes automated tests to save our mental health 😜
```
Comment (jancama2) [inline pages/Testing.md]: Typo => `Here come`

File: pages/testing/Essentials.md [hunk 3]
```diff
@@ -0,0 +1,399 @@
+# Testing Essentials
+
+For testing generally, you will always need at least **testing framework** & usually also **mocking library**.
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `In general, you will ...`

File: pages/testing/Essentials.md [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+  * [Comparation helpers](#comparation-helpers)
+* [Mocking](#mocking)
+  * [Basic mocks](#basic-mocks)
+  * [Timer mocks](#timer-mocks)
+
+---
+
+## Matchers
+
+> Don't worry about mocking sometimes used in examples, we'll discuss it in next section 😉
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `Don't worry about the mocking I use in examples, ...'

File: pages/testing/Essentials.md [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+
+What is it? Long story short: "function that test expected value against real one". They make subject of testing more obvious.
```
[the remark this answers is not in the document]
Comment (jancama2) [reply pages/testing/Essentials.md]: `an expected value`, `a real one`

File: pages/testing/Essentials.md [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+
+There is [big bunch of matchers](https://jestjs.io/docs/en/expect#methods), but the ones most used in tests are `toBe`, `toEqual`, `toHaveBeenCalledTimes` and `toHaveBeenCalledWith`. Let's review theme one by one and look at their usage and props/cons.
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `the ones usually used` ?

Comment (jancama2) [reply pages/testing/Essentials.md]: `and take a look at ...` ?

File: pages/testing/Essentials.md [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+    const game = { done: true }
+    
+    expect(game.done).toBe(true) // passes
+    expect(game).toBe({ done: true }) // fails
+})
+```
+
+### `.toEqual(value)`
+
+* Does deep compare so it's useful for comparation complex objects
```
[the remark this answers is not in the document]
Comment (jancama2) [reply pages/testing/Essentials.md]: `Does a deep comparison.. ` ?

File: pages/testing/Essentials.md [hunk 8]
```diff
@@ -0,0 +1,399 @@
...
+    ]
+    
+    expect(players).toHaveLength(2) // passes
+    expect(players.length).toEqual(2) // also passes
+})
+```
+
+### `.toHaveBeenCalledTimes(count)`
+
+* Test if mock function was called n-times exactly
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `Tests if a mock function ... `

File: pages/testing/Essentials.md [hunk 9]
```diff
@@ -0,0 +1,399 @@
...
+     
+    const user = getFirebaseUser()
+
+    expect(firebase.auth).toHaveBeenCalledTimes(1)
+})
+```
+
+### `.toHaveBeenCalledWith(arg1, arg2, ...)`
+
+* Test what arguments was function called with
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `Tests what arguments was a function called with ... `

File: pages/testing/Essentials.md [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+    expect(firebase.auth.GoogleAuthProvider.credential).toHaveBeenCalledWith(token)
+    expect(firebase.auth.signInWithCredential).toHaveBeenCalledWith('fake credential')
+})
+```
+
+### Comparation helpers
+
+There some comparation helpers that easy your work when comparing complex objects, or want to check only particular arguments of a function call.
+
+Most of the helpers is negatable (usually those which makes sense to negate) by prepending `.not`, eg. `expect.not.arrayContaining`.
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `Most of the helpers are... `

File: pages/testing/Essentials.md [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+    expect(point).toEqual(expect.objectContaining({
+        x: expect.any(Number),
+        y: expect.any(Number),
+    }))
+})
+```
+
+#### `expect.stringContaining(string)` & `expect.stringMatching(string|regexp)`
+
+You wouldn't use these helpers for simple value comparation, cause there are `toContain` and `toMatch` matchers, that are easier and more suitable to use. But you will find them handy in combination with `.arrayCotaining` and `.objectContaining`
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `comparison, because ... `

File: pages/testing/Essentials.md [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+
+---
+
+That's  all about matchers for now, don't forget to discover all of them at the [documentation page](https://jestjs.io/docs/en/expect#methods). There are plenty of other matchers that you might find useful like `.toContain`, `.toBeNull`, `.toThrow`, `.toMatch`, `toHaveLength`, etc. but the described ones are basics.
+
+And one more thing - you can extend basic set of matchers by custom ones when you need some, eg. for some specific environments like [DOM](https://github.com/testing-library/jest-dom) or [React Native](https://github.com/testing-library/jest-native).
+
+## Mocking
+
+>Technique that allows you to focus on testing one specific part of code, faking other parts assuming they work as intended.
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `A technique allowing you to ... `

File: pages/testing/Essentials.md [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+    * [callbacks](https://jestjs.io/docs/en/asynchronous#callbacks)
+    * [promises](https://jestjs.io/docs/en/asynchronous#promises)
+    * [async/await](https://jestjs.io/docs/en/asynchronous#async-await)
+
+
+### Why we need mocking?
+
+* **SPEED**‼ Less real code to execute means faster tests run.
+* **Focus** When dependencies are mocked and only tested code is run, we can quickly identify why test fails.
+* Of course, there is more reasons for mocking but I am not able to list all of them here. You can find more of them in Kent C. Doods [article about javascript mocks](#resources)
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `there are more reasons ...`

File: pages/testing/Essentials.md [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+
+### Basic mocks
+
+We will ilustrate it on mocking 3rd party library, but you can use it same way also for other sources, modules, etc. in app.
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `the same way ...`

File: pages/testing/Essentials.md [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+        routingSelector: jest.fn(),
+    }
+})
+```
+
+You can also check [module examples in jest repository](https://github.com/facebook/jest/tree/master/examples/module-mock) as well as [manual mock examples](https://github.com/facebook/jest/tree/master/examples/manual-mocks).
+
+### Timer mocks
+
+Sometimes you code depends on some timing, maybe you use `setTimeout` or other kind of delaying code by time, and you would like to conrol the timing for tests.
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `Sometimes your ... `

File: pages/testing/Essentials.md [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+})
+```
+
+You can also check [module examples in jest repository](https://github.com/facebook/jest/tree/master/examples/module-mock) as well as [manual mock examples](https://github.com/facebook/jest/tree/master/examples/manual-mocks).
+
+### Timer mocks
+
+Sometimes you code depends on some timing, maybe you use `setTimeout` or other kind of delaying code by time, and you would like to conrol the timing for tests. 
+
+That easily solved in Jest by faking timers with [`jest.useFakeTimers`](https://jestjs.io/docs/en/timer-mocks).
```
Comment (jancama2) [inline pages/testing/Essentials.md]: `That is easily solved ... `

File: pages/testing/Essentials.md (hunk 4 above)
Comment (horaklukas) [reply pages/testing/Essentials.md]: I don't want to write it from a first person view.

**Pull request merged**